File: db_beam_elixir/db_beam/lib/ds_processes.ex
defmodule DsProcesses do
  @moduledoc false



  def f() do

    receive do
      {:msg, msg} -> Blogs.log_info("Got message: #{msg}")
                     f()
      {:close} -> Blogs.log_info("Good bye")

    end

  end


  def monitor_f(pid) do
    ref = Process.monitor(pid)
    Blogs.log_info("Monitor Ref: #{inspect(ref)}")
    receive do
      {:DOWN, monitor_reference, :process, pid, reason} ->
        Blogs.log_info(
          "I am the monitor for pid: #{inspect(pid)}, pid down reason:#{inspect(reason)}, ref #{
            inspect(inspect(monitor_reference))
          } "
        )
    end

  end






  # won't delete the chain, but handle the Exit flag.
  def sup_f() do
    Process.flag(:trap_exit, true) # %% can apply more than one time, it is always one
    pid = spawn_link(&f/0)
    pid_monitor = spawn(DsProcesses, :monitor_f, [pid])
    Blogs.log_info(
      "TEE - The sup PID is: #{inspect(self())}, pid f is #{inspect(pid)},  pid monitor:#{inspect(pid_monitor)}"
    )

    receive do
      {:EXIT, b, reason} ->
        Blogs.log_info("TEE - Received 'EXIT' message pid:#{inspect(b)} Reason:#{inspect(reason)}")
        case reason do
          :normal -> Blogs.log_info("TEE - Stopped normally")
          value -> Blogs.log_info("TEE - Unexpected Stop #{value} going to restart it ~n")
                   sup_f() #%% This is for restart
        end

      value -> Blogs.log_info("sup_f :#{inspect(value)}")
    end
  end



  # send(pid("0.116.0"), {:msg, "ciao"})
  # try with exit(PID, normal)
  # and with exit(PID, my_error)
  def spawn_sup_f() do
    spawn(&sup_f/0)
  end







  # diffent between process flag and not, chain delete
  #try to kill the super here to see the chain delete
  def sup_f_no() do
    pid = spawn_link(&f/0)
    pid_monitor = spawn(DsProcesses, :monitor_f, [pid])
    Blogs.log_info(
      "TED - The sup PID is: #{inspect(self())} and the f is PIDF:#{inspect(pid)} PID_Monitor:#{inspect(pid_monitor)}"
    ) # note the self() is always the same
    receive do
      value -> Blogs.log_info("sup_f_no #{inspect(value)}")
    end

  end

  def spawn_sup_f_no() do
    spawn(&sup_f_no/0)
  end



  def spawn_with_dictonary() do
    :proc_lib.spawn(&f/0)
  end


end

File: dp_beam/src/ds_processes.erl
%%%-------------------------------------------------------------------
%%% @author gabriele
%%% @copyright (C) 2017, <COMPANY>
%%% @doc
%%%
%%% @end
%%% Created : 06. Dec 2017 09:27
%%%-------------------------------------------------------------------
-module(ds_processes).
-author("gabriele").

%% API
-export([f/0, sup_f/0, spawn_sup_f/0, sup_f_no/0, monitor_f/1, spawn_sup_f_no/0, spawn_with_dictonary/0]).


f() -> receive
           {msg, MSG} -> io:format("Got message:~p~n", [MSG]), f();
           {close} -> io:format("Good bye", [])
       end.

%% won't delete the chain, but handle the Exit flag.
sup_f() ->
    process_flag(trap_exit, true), %% can apply more than one time, it is always one
    PID = spawn_link(fun f/0),
    PID_Monitor = spawn(ds_processes, monitor_f, [PID]),
    io:format("TEE - The sup PID is: ~p and the f is PIDF:~p PID_Monitor:~p ~n", [self(), PID, PID_Monitor]), %% note the self() is always the same
    receive
        {'EXIT', B, Reason} ->
            io:format("TEE - Received 'EXIT' Message PID:~p Reason:~p~n", [B, Reason]),

            case Reason of
                normal -> io:format("TEE - Stopped normally ~n", []);
                Value ->
                    io:format("TEE - Unexpected Stop ~p going to restart it ~n", [Value]),
                    sup_f() %% This is for restart
            end;

        Value -> io:format("sup_f ~p", [Value])
    end.


%% try with exit(PID, normal)
%% and with exit(PID, my_error)
%% for example exit(list_to_pid("<0.650.0>"), my_errror).
%% The "The sup PID" is always the same 'cause of trap_exit
%% use always process info to check the status and show the links/mornitor flag
%% process_info(list_to_pid( SPANNED_PID ),monitored_by). who is monitor
%% process_info(list_to_pid( PID_Monitor  ),monitors). who is monitoring

spawn_sup_f() -> spawn(fun sup_f/0).





%% diffent between process flag and not, chain delete
%% try to kill the super here to see the chain delete
sup_f_no() ->
    PID = spawn_link(fun f/0),
    PID_Monitor = spawn(ds_processes, monitor_f, [PID]),
    io:format("TED - The sup PID is: ~p and the f is PIDF:~p PID_Monitor:~p ~n", [self(), PID, PID_Monitor]), %% note the self() is always the same
    receive
        Value -> io:format("sup_f_no ~p", [Value])
    end.



spawn_sup_f_no() -> spawn(fun sup_f_no/0).

monitor_f(Pid) ->
    Ref = erlang:monitor(process, Pid),
    io:format("Monitor Ref:~p ~n", [Ref]),
    receive
        {'DOWN', MonitorReference, process, Pid, Reason} ->
            io:format("I am a monitor PID:~p Reason:~p  MonitorReference:~p ~n", [Pid, Reason, MonitorReference])
    end.


spawn_with_dictonary() -> proc_lib:spawn(fun f/0).

